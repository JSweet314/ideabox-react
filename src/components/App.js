import React, { Component } from 'react';
import '../App.css';
import IdeaForm from '../containers/IdeaForm';
import CardContainer from '../containers/CardContainer';

class App extends Component {
  constructor() {
    super();
    this.state = {
      ideas: [],
      visibleIdeas: []
    };

    this.handleNewIdea = this.handleNewIdea.bind(this);
    this.updateIdeaQuality = this.updateIdeaQuality.bind(this);
    this.removeIdea = this.removeIdea.bind(this);
    this.searchIdeas = this.searchIdeas.bind(this);
  }

  componentDidMount() {
    let ideas = [];
    let previousBox = localStorage.getItem('ideabox');

    if (previousBox) {
      ideas = JSON.parse(previousBox);
    }
    
    this.setState({
      ideas,
      visibleIdeas: ideas
    });
  }

  handleNewIdea(idea) {
    const ideas = [idea, ...this.state.ideas];

    localStorage.setItem(
      'ideabox',
      JSON.stringify(ideas)
    );
    this.setState({
      ideas,
      visibleIdeas: ideas
    });
  }

  removeIdea(e, id) {
    e.preventDefault();
    const filteredIdeas = this.state.ideas.filter(idea => {
      return idea.id !== id;
    });

    localStorage.setItem('ideabox', JSON.stringify(filteredIdeas));

    this.setState({
      ideas: filteredIdeas,
      visibleIdeas: filteredIdeas
    });
  }

  updateIdeaQuality(event, id) {
    const {idea, index} = this.state.ideas.reduce((obj, idea, index) => {
      if (idea.id === id) {
        obj = {idea, index};
      }
      return obj;
    }, {});
  
    switch (event.target.className) {
    case 'upVoteBtn':
      if (idea.quality === 'swill') {
        idea.quality = 'plausible';
      } else {
        idea.quality = 'genius';
      }
      break;        
    default:
      if (idea.quality === 'genius') {
        idea.quality = 'plausible';
      } else {
        idea.quality = 'swill';
      }
      break;
    }
    
    const ideas = [
      ...this.state.ideas.slice(0, index),
      idea,
      ...this.state.ideas.slice(index + 1)
    ];

    localStorage.setItem('ideabox', JSON.stringify(ideas));

    this.setState({
      ideas,
      visibleIdeas: [...this.state.visibleIdeas]
    });
  }

  searchIdeas(searchValue) {
    let results = this.state.ideas.filter(idea => {
      return idea.title.includes(searchValue) 
        || idea.body.includes(searchValue);
    });

    this.setState({
      visibleIdeas: results
    });
  }

  render() {
    return (
      <div className="app">
        <IdeaForm 
          handleNewIdea={this.handleNewIdea} 
        />
        <CardContainer 
          visibleIdeas={this.state.visibleIdeas}
          updateIdeaQuality={this.updateIdeaQuality}
          removeIdea={this.removeIdea}
          searchIdeas={this.searchIdeas}
        />
      </div>
    );
  }
}

export default App;