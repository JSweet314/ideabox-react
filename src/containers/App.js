import React, { Component } from 'react';
import IdeaForm from './IdeaForm.js';
import IdeaCards from '../components/IdeaCards.js';
import '../styles/App.css';

class App extends Component {
  constructor() {
    super();
    this.state = {
      ideas: [],
      visibleIdeas: []
    };
    this.handleNewIdea = this.handleNewIdea.bind(this);
    this.updateIdeaQuality = this.updateIdeaQuality.bind(this);
    this.removeIdea = this.removeIdea.bind(this);
    this.searchIdeas = this.searchIdeas.bind(this);
    this.filterByQuality = this.filterByQuality.bind(this);
  }

  componentDidMount() {
    let ideas = [];
    let previousBox = localStorage.getItem('ideabox');

    if (previousBox) {
      ideas = JSON.parse(previousBox);
    }
    
    this.setState({ideas, visibleIdeas: ideas});
  }

  handleNewIdea(idea) {
    const ideas = [idea, ...this.state.ideas];

    localStorage.setItem('ideabox', JSON.stringify(ideas));
    this.setState({ideas, visibleIdeas: ideas});
  }

  removeIdea(id) {
    const filteredIdeas = this.state.ideas.filter(idea => {
      return idea.id !== id;
    });

    localStorage.setItem('ideabox', JSON.stringify(filteredIdeas));

    this.setState({ideas: filteredIdeas, visibleIdeas: filteredIdeas});
  }

  updateIdeaQuality(event, id) {
    const {idea, index} = this.state.ideas.reduce((data, idea, index) => {
      if (idea.id === id) {
        data = {idea, index};
      }
      return data;
    }, {});
  
    switch (event.target.className) {
    case 'upVoteBtn':
      idea.quality = idea.quality === 'swill' ? 'plausible' : 'genius';
      break;        
    default:
      idea.quality = idea.quality === 'genius' ? 'plausible' : 'swill';
      break;
    }
    
    const ideas = [
      ...this.state.ideas.slice(0, index),
      idea,
      ...this.state.ideas.slice(index + 1)
    ];

    localStorage.setItem('ideabox', JSON.stringify(ideas));

    this.setState({ideas});
  }

  searchIdeas(searchValue) {
    let visibleIdeas = this.state.ideas.filter(idea => {
      return (
        idea.title.includes(searchValue) || idea.body.includes(searchValue)
      );
    });
    
    this.setState({visibleIdeas});
  }

  filterByQuality(qualityFilter) {
    let visibleIdeas;

    if (qualityFilter === 'all') {
      visibleIdeas = this.state.ideas;
    } else {
      visibleIdeas = this.state.ideas.filter(idea => {
        return idea.quality === qualityFilter;
      });
    }

    this.setState({visibleIdeas, qualityFilter});
  }

  render() {
    return (
      <div className="app">
        <IdeaForm handleNewIdea={this.handleNewIdea} />
        <IdeaCards 
          visibleIdeas={this.state.visibleIdeas}
          updateIdeaQuality={this.updateIdeaQuality}
          removeIdea={this.removeIdea}
          searchIdeas={this.searchIdeas}
          filterByQuality={this.filterByQuality}
        />
      </div>
    );
  }
}

export default App;