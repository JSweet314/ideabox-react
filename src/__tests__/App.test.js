import React from 'react';
import ReactDOM from 'react-dom';
import {shallow, mount} from 'enzyme';
import App from '../containers/App.js';

describe('app', () => {
  let app, event, idea1, idea2;

  beforeEach(() => {
    event = {
      target: {
        className: 'upVoteBtn'
      }
    };

    idea1 = {
      id: 1234,
      title: 'Hello',
      body: 'World!',
      quality: 'plausible'
    };

    idea2 = {
      id: 4321,
      title: 'New App!',
      body: 'Woo Hoo',
      quality: 'genius'
    };

    localStorage.clear();

    app = shallow(<App />);
  });

  it('should render our good friend, App', () => {
    const div = document.createElement('div');

    ReactDOM.render(<App />, div);
  });

  it('should render an IdeaForm and a CardContainer', () => {
    expect(app).toMatchSnapshot();
  });  

  it('should render the whole application', () => {
    app = mount(<App />);
    expect(app).toMatchSnapshot();
  });

  it('should hold an array of ideas in state', () => {
    expect(app.state('ideas')).toEqual([]);
  });

  it('should be able to add a new idea to its array of ideas', () => {
    expect(app.state('ideas')).toEqual([]);
    
    app.instance().handleNewIdea(idea1);

    expect(app.state('ideas')).toEqual([idea1]);
  });

  it('should allow for multiple ideas', () => {
    expect(app.state('ideas')).toEqual([]);

    app.instance().handleNewIdea(idea1);

    expect(app.state('ideas')).toEqual([idea1]);

    app.instance().handleNewIdea(idea2);

    expect(app.state('ideas')).toEqual([idea2, idea1]);
  });

  it('should store ideas in localStorage for latter', () => {
    expect(app.state('ideas')).toEqual([]);

    app.instance().handleNewIdea(idea1);

    expect(app.state('ideas')).toEqual([idea1]);
    expect(JSON.parse(localStorage.getItem('ideabox')).length).toEqual(1);

    app.instance().handleNewIdea(idea2);

    expect(app.state('ideas')).toEqual([idea2, idea1]);
    expect(JSON.parse(localStorage.getItem('ideabox')).length).toEqual(2);
  });

  it('should be able to remove an idea from its array', () => {
    app.instance().handleNewIdea(idea1);

    expect(app.state('ideas')).toEqual([idea1]);
    
    app.instance().removeIdea(1234);

    expect(app.state('ideas')).toEqual([]);
  });

  it('should allow the user to update the quality of an idea in the ideas array', () => {
    app.instance().handleNewIdea(idea1);

    expect(app.state('ideas')[0].quality).toEqual('plausible');

    app.instance().updateIdeaQuality(event, 1234);

    expect(app.state('ideas')[0].quality).toEqual('genius');

    event.target.className = 'downVoteBtn';
    app.instance().updateIdeaQuality(event, 1234);

    expect(app.state('ideas')[0].quality).toEqual('plausible');
  });

  it('should be able to search through and display specific ideas', () => {
  });

  it('should be able to display ideas by quality', () => {
  });

  it('should retrieve ideas from localStorage if available', () => {
    expect(app.state('ideas')).toEqual([]);
    
    localStorage.setItem('ideabox', JSON.stringify([idea2, idea1]));
    app = shallow(<App />);

    expect(app.state('ideas')).toEqual([idea2, idea1]);
  });
});