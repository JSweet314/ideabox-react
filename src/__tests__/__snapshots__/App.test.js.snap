// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app should render an IdeaForm and a CardContainer 1`] = `
<div
  className="app"
>
  <IdeaForm
    handleNewIdea={[Function]}
  />
  <QualityFilter
    filterByQuality={[Function]}
    searchIdeas={[Function]}
  />
  <IdeaCards
    ideas={Array []}
    qualityFilter=""
    removeIdea={[Function]}
    searchValue=""
    updateIdeaQuality={[Function]}
  />
</div>
`;

exports[`app should render the whole application 1`] = `
<App>
  <div
    className="app"
  >
    <IdeaForm
      handleNewIdea={[Function]}
    >
      <form
        onSubmit={[Function]}
      >
        <h1>
          <span>
            idea
          </span>
          box
        </h1>
        <input
          aria-label="idea title input"
          className="titleInput"
          maxLength="120"
          name="title"
          onChange={[Function]}
          placeholder="Idea Title"
          type="text"
          value=""
        />
        <p>
          Characters remaining: 
          120
        </p>
        <input
          aria-label="idea body input"
          className="bodyInput"
          maxLength="120"
          name="body"
          onChange={[Function]}
          placeholder="Idea Body"
          type="text"
          value=""
        />
        <p>
          Characters remaining: 
          120
        </p>
        <button
          disabled={true}
          type="submit"
        >
          save
        </button>
      </form>
    </IdeaForm>
    <QualityFilter
      filterByQuality={[Function]}
      searchIdeas={[Function]}
    >
      <div
        className="quality-filter"
      >
        <input
          className="search"
          onChange={[Function]}
          placeholder="SEARCH"
          type="search"
        />
        <label
          htmlFor=""
        >
          <span>
            Filter by Idea Quality: 
          </span>
          <select
            className="quality-filter__select"
            name="quality"
            onChange={[Function]}
          >
            <option
              value="all"
            >
              all
            </option>
            <option
              value="swill"
            >
              swill
            </option>
            <option
              value="plausible"
            >
              plausible
            </option>
            <option
              value="genius"
            >
              genius
            </option>
          </select>
        </label>
      </div>
    </QualityFilter>
    <IdeaCards
      ideas={Array []}
      qualityFilter=""
      removeIdea={[Function]}
      searchValue=""
      updateIdeaQuality={[Function]}
    >
      <div
        className="card-container"
      />
    </IdeaCards>
  </div>
</App>
`;
